### PR TITLE
(PC-27868) feat(ChangeEmail): Add banner when old email change is disabled

### PR DESCRIPTION
**Pull request by aliraiki**
Link to JIRA ticket: https://passculture.atlassian.net/browse/PC-27868

## Checklist

I have:

- [x] Made sure my feature is working on the relevant real / virtual devices (native and web).
- [x] Written **unit tests** native (and web when implementation is different) for my feature.
- [x] Added a **screenshot** for UI tickets or deleted the screenshot section if no UI change
- [ ] If my PR is a bugfix, I add the link of the "résolution de problème sur le bug" [on Notion][1]

## Screenshots
| Platform         | Mockup/Before | After |
| :--------------- | :-----------: | :---: |
| iOS              |Simulator Screenshot - iPhone 13 - 2024-02-13 at 12 00 19     Simulator Screenshot - iPhone 13 - 2024-02-13 at 11 45 13          |    Simulator Screenshot - iPhone 13 - 2024-02-13 at 11 43 29 Simulator Screenshot - iPhone 13 - 2024-02-13 at 11 44 32  |
| Desktop - Chrome | Capture d’écran 2024-02-13 à 11 49 48  Capture d’écran 2024-02-13 à 11 22 21     |  Capture d’écran 2024-02-13 à 11 00 41  Capture d’écran 2024-02-13 à 11 22 39      |

[1]: https://www.notion.so/passcultureapp/R-solution-de-probl-mes-sur-les-bugs-5dd6df8f6a754e6887066cf613467d0a

### CODE REVIEW

File: src/libs/firebase/firestore/featureFlags/useFeatureFlag.ts [hunk 1]
```diff
@@ -6,18 +6,32 @@ import { getAppBuildVersion } from 'libs/packageJson'
...
 // firestore feature flag documentation :
 // https://www.notion.so/passcultureapp/Feature-Flag-e7b0da7946f64020b8403e3581b4ed42#fff5fb17737240c9996c432117acacd8
-export const useFeatureFlag = (featureFlag: RemoteStoreFeatureFlags): boolean | undefined => {
-  const [minimalBuildNumber, setMinimalBuildNumber] = useState<number | null>()
+export const useFeatureFlag = (
+  remoteStorefeatureFlag: RemoteStoreFeatureFlags
+): boolean | undefined => {
+  const [buildNumberConfig, setBuildNumberConfig] = useState<{
+    minimalBuildNumber?: number | null
+    maximalBuildNumber?: number | null
```
Comment (bruno-ebstein-pass-culture) [inline src/libs/firebase/firestore/featureFlags/useFeatureFlag.ts:14]: les valeurs étant encapsulés, je ne vois pas l'intéret d'avoir 2 types pour exprimer l'absence de valeur

```suggestion
    minimalBuildNumber?: number
    maximalBuildNumber?: number
```

Comment (aliraiki) [reply src/libs/firebase/firestore/featureFlags/useFeatureFlag.ts:14]: Le `undefined` c'est pour dire qu'on n'a toujours pas la réponse de firestore, le `null` c'est quand aucune valeur n'est précisé sur firestore

File: TODO.md [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+| version courante supérieure à la version minimale requise    |        2         |    1    |         | activé                 |
+| version courante égale à la version minimale requise         |        1         |    1    |         | activé                 |
+| version courante inférieure à la version minimale requise    |        1         |    2    |         | désactivé              |
+| version courante égale à la version maximale requise         |        1         |         |    1    | activé                 |
+| version courante supérieure à la version maximale requise    |        2         |         |    1    | désactivé              |
+| version courante inférieure à la version maximale requise    |        1         |         |    2    | activé                 |
+| version courante entre version minimale et maximale requises |        2         |    1    |    3    | activé                 |
+| version minimale égale à la version maximale                 |        1         |    1    |    1    | désactivé              |
+| version minimale supérieure à la version maximale            |        2         |    3    |    1    | désactivé + log sentry |
```
Comment (bruno-ebstein-pass-culture) [inline TODO.md:42]: je n'ai pas vu le log dans le code

File: src/libs/firebase/firestore/featureFlags/useFeatureFlag.ts (hunk 1 above)
Comment (bruno-ebstein-pass-culture) [reply src/libs/firebase/firestore/featureFlags/useFeatureFlag.ts:14]: en cas d'égalité, on veut que ça soit activé, donc on pourrait carrement avoir

```suggestion
    minimalBuildNumber: number
    maximalBuildNumber: number
```

si pas de version, fallback sur la version courante

je ne sais pas si c'est une bonne idée

File: src/libs/firebase/firestore/featureFlags/useFeatureFlag.native.test.ts [hunk 3]
```diff
@@ -64,4 +65,38 @@ describe.each([
       expect(result.current).toBe(firebaseFeatureFlag)
     }
   )
+
+  it('should deactivate FF when no build number is given', async () => {
```
Comment (bruno-ebstein-pass-culture) [inline src/libs/firebase/firestore/featureFlags/useFeatureFlag.native.test.ts:69]: les `it` qui suivent devraient probablement etre dans un autre `describe`

on avait que les tests actuels étaient améliorables, on a choisi de ne pas mettre le focus dessus dans un premier temps

itérer sur les FF pour ces tests n'apporte rien

File: src/libs/firebase/firestore/featureFlags/getFeatureFlag.ts [hunk 4]
```diff
@@ -5,13 +5,14 @@ import { captureMonitoringError } from 'libs/monitoring'
...
-): Promise<void | { minimalBuildNumber: number } | null> =>
+): Promise<void | { minimalBuildNumber?: number; maximalBuildNumber?: number } | null> =>
   firestoreRemoteStore
     .collection(RemoteStoreCollections.FEATURE_FLAGS)
     .doc(env.ENV)
     .get()
     .then((docSnapshot) => {
-      const firebaseFeatureFlag = docSnapshot.get<{ minimalBuildNumber: number }>(featureFlag)
+      const firebaseFeatureFlag =
+        docSnapshot.get<{ minimalBuildNumber: number; maximalBuildNumber?: number }>(featureFlag)
```
Comment (bruno-ebstein-pass-culture) [inline src/libs/firebase/firestore/featureFlags/getFeatureFlag.ts:15]: c'est étonnant que TypeScript ne rale pas

```suggestion
        docSnapshot.get<{ minimalBuildNumber?: number; maximalBuildNumber?: number }>(featureFlag)
```

on n'a pas eu le temps de finir de l'écrire, hier

File: src/features/profile/pages/TrackEmailChange/TrackEmailChange.native.test.tsx [hunk 5]
```diff
@@ -281,4 +296,41 @@ describe('TrackEmailChange', () => {
...
+          status,
+        },
+        isLoading: false,
+      } as UseEmailUpdateStatusMock)
+      render(<TrackEmailChange />)
+
+      const stepCards = screen.getAllByTestId('stepcard-container')
+
+      stepCards.forEach((stepCard) => {
+        expect(stepCard.props.type).toBe('disabled')
```
Comment (bruno-ebstein-pass-culture) [inline src/features/profile/pages/TrackEmailChange/TrackEmailChange.native.test.tsx:332]: je suis étonné qu'on soit obligé de vérifier les props, ça me parait un peu white box

est-ce qu'on aurait une assertion du style `.toBeDisabled()` ?

File: TODO.md [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+| aucune version minimale ou maximale spécifiée                |        1         |         |         | désactivé              |
+| version courante supérieure à la version minimale requise    |        2         |    1    |         | activé                 |
+| version courante égale à la version minimale requise         |        1         |    1    |         | activé                 |
+| version courante inférieure à la version minimale requise    |        1         |    2    |         | désactivé              |
+| version courante égale à la version maximale requise         |        1         |         |    1    | activé                 |
+| version courante supérieure à la version maximale requise    |        2         |         |    1    | désactivé              |
+| version courante inférieure à la version maximale requise    |        1         |         |    2    | activé                 |
+| version courante entre version minimale et maximale requises |        2         |    1    |    3    | activé                 |
+| version minimale égale à la version maximale                 |        1         |    1    |    1    | désactivé              |
```
Comment (bruno-ebstein-pass-culture) [inline TODO.md:41]: pourquoi on s'était dit que c'était `désactivé` ?

File: src/libs/firebase/firestore/featureFlags/useFeatureFlag.ts (hunk 1 above)
Comment (bruno-ebstein-pass-culture) [reply src/libs/firebase/firestore/featureFlags/useFeatureFlag.ts:14]: on n'expose pas cette information hors du hook

on a le meme comportement entre "on n'a pas encore l'info" et "on n'a pas l'info"

est-ce qu'il y a une utilité à les séparer ?

File: src/features/profile/pages/TrackEmailChange/TrackEmailChange.native.test.tsx (hunk 5 above)
Comment (aliraiki) [reply src/features/profile/pages/TrackEmailChange/TrackEmailChange.native.test.tsx:332]: L'élément n'est pas un bouton ou un élément avec lequel on peut interagir (c'est juste une view) donc on ne peut pas utiliser `toBeDisabled`

File: TODO.md (hunk 6 above)
Comment (aliraiki) [reply TODO.md:41]: Je ne sais pas trop vu que j'étais pas au MOB, je laisse @Lucasbeneston répondre si possible

File: src/features/profile/pages/TrackEmailChange/TrackEmailChange.native.test.tsx (hunk 5 above)
Comment (bruno-ebstein-pass-culture) [reply src/features/profile/pages/TrackEmailChange/TrackEmailChange.native.test.tsx:332]: dommage

File: TODO.md (hunk 6 above)
Comment (Lucasbeneston) [reply TODO.md:41]: Mmmh je ne sais plus... en vrai c'est une question pour le produit ? Victor et Mouna ?

File: src/libs/firebase/firestore/featureFlags/useFeatureFlag.ts (hunk 1 above)
Comment (bruno-ebstein-pass-culture) [reply src/libs/firebase/firestore/featureFlags/useFeatureFlag.ts:14]: proposition (pas testé, écrit à l'arrache)

```ts
export const useFeatureFlag = (remoteStorefeatureFlag: RemoteStoreFeatureFlags): boolean => {
  const appBuildVersion = getAppBuildVersion()
  const [{ minimalBuildNumber, maximalBuildNumber }, setBuildNumberConfig] = useState<{
    minimalBuildNumber: number
    maximalBuildNumber: number
  }>({
    minimalBuildNumber: appBuildVersion,
    maximalBuildNumber: appBuildVersion,
  })

  useEffect(() => {
    async function updateFeatureFlag() {
      const featureFlag = await getFeatureFlag(remoteStorefeatureFlag)
      setBuildNumberConfig({
        minimalBuildNumber: featureFlag?.minimalBuildNumber || appBuildVersion,
        maximalBuildNumber: featureFlag?.maximalBuildNumber || appBuildVersion,
      })
    }

    updateFeatureFlag()
  }, [appBuildVersion, remoteStorefeatureFlag])

  return minimalBuildNumber <= appBuildVersion && appBuildVersion <= maximalBuildNumber
}
```

Comment (bruno-ebstein-pass-culture) [reply src/libs/firebase/firestore/featureFlags/useFeatureFlag.ts:14]: j'ai oublié le cas où il faut lever une erreur

```ts
if (minimalBuildNumber > maximalBuildNumber) error()
```

File: src/libs/firebase/firestore/featureFlags/getFeatureFlag.ts (hunk 4 above)
Comment (aliraiki) [reply src/libs/firebase/firestore/featureFlags/getFeatureFlag.ts]: J'ai refacto pour réutiliser ce type dans le retour de la fonction et le `get`

File: src/libs/firebase/firestore/featureFlags/useFeatureFlag.native.test.ts (hunk 3 above)
Comment (aliraiki) [reply src/libs/firebase/firestore/featureFlags/useFeatureFlag.native.test.ts:29]: J'ai ajouté un commit pour refacto les tests

File: TODO.md (hunk 2 above)
Comment (aliraiki) [reply TODO.md]: Je viens de lire cette partie de la todo en mettant à jour la doc Notion, j'ai ajouté la gestion de ces edge cases

File: src/libs/firebase/firestore/featureFlags/useFeatureFlag.ts (hunk 1 above)
Comment (aliraiki) [reply src/libs/firebase/firestore/featureFlags/useFeatureFlag.ts:15]: Avec cette implém, les cas suivants rentrent en conflit : 
- Si minimal = maximal, on veut désactiver le FF
- Si maximal ou minimal = version actuel, on veut activer le FF

Comment (bruno-ebstein-pass-culture) [reply src/libs/firebase/firestore/featureFlags/useFeatureFlag.ts:15]: > Si minimal = maximal, on veut désactiver le FF

pour quelles raisons voudrait-on faire ça ?

si on veut activer un FF seulement sur une version

---

> Si maximal ou minimal = version actuel, on veut activer le FF

avec le code que j'ai proposé, j'ai l'impression que c'est le cas

Comment (bruno-ebstein-pass-culture) [reply src/libs/firebase/firestore/featureFlags/useFeatureFlag.ts:15]: et si c'est pas le cas, ça me semble plus facile d'expliciter avec une condition en amont qui montre ce edge case

File: TODO.md (hunk 6 above)
Comment (bruno-ebstein-pass-culture) [reply TODO.md]: si on veut activer un FF seulement sur une version, ça pourrait etre pratique

File: src/libs/firebase/firestore/featureFlags/useFeatureFlag.ts (hunk 1 above)
Comment (aliraiki) [reply src/libs/firebase/firestore/featureFlags/useFeatureFlag.ts:15]: > pour quelles raisons voudrait-on faire ça ?
C'était marqué dans la TODO

> avec le code que j'ai proposé, j'ai l'impression que c'est le cas
Oui et non, par exemple si sur firestore on met minimal = version actuelle, et qu'on ne précise pas de version maximale, le FF sera désactivé (parce que la valeur par défaut de maximal sera version actuelle, du coup on a minimal = maximal, et dans ce cas on désactive le FF d'après la TODO)

File: TODO.md (hunk 6 above)
Comment (bruno-ebstein-pass-culture) [reply TODO.md]: après réflexion (unilatérale), je pense qu'il faudrait que ça soit activé

Comment (aliraiki) [reply TODO.md]: Je pense aussi, d'après le tableau si version actuelle = version minimale ou version actuelle = version maximale on active le FF, du coup je pense que si la version actuelle = version minimale = version maximale on devrait activer aussi pour rester cohérent

File: src/libs/firebase/firestore/featureFlags/useFeatureFlag.ts [hunk 7]
```diff
@@ -2,22 +2,54 @@ import { useEffect, useState } from 'react'
...
 import { getFeatureFlag } from 'libs/firebase/firestore/featureFlags/getFeatureFlag'
 import { RemoteStoreFeatureFlags } from 'libs/firebase/firestore/types'
+import { eventMonitoring } from 'libs/monitoring'
 import { getAppBuildVersion } from 'libs/packageJson'
 
 // firestore feature flag documentation :
 // https://www.notion.so/passcultureapp/Feature-Flag-e7b0da7946f64020b8403e3581b4ed42#fff5fb17737240c9996c432117acacd8
-export const useFeatureFlag = (featureFlag: RemoteStoreFeatureFlags): boolean | undefined => {
-  const [minimalBuildNumber, setMinimalBuildNumber] = useState<number | null>()
+export const useFeatureFlag = (
```
Comment (bruno-ebstein-pass-culture) [inline src/libs/firebase/firestore/featureFlags/useFeatureFlag.ts:10]: la fonction a une grosse compléxitée

Comment (aliraiki) [reply src/libs/firebase/firestore/featureFlags/useFeatureFlag.ts:10]: Je suis d'accord, je vais refacto maintenant qu'on sait quand est-ce qu'on veut activer ou désactiver dans le cas minimale = maximale

File: src/libs/firebase/firestore/featureFlags/useFeatureFlag.ts [hunk 8]
```diff
@@ -2,22 +2,52 @@ import { useEffect, useState } from 'react'
...
+      setBuildNumberConfig({
+        minimalBuildNumber: featureFlag.minimalBuildNumber,
+        maximalBuildNumber: featureFlag.maximalBuildNumber,
+      })
+    })
+  }, [remoteStorefeatureFlag])
+
+  const { minimalBuildNumber, maximalBuildNumber } = buildNumberConfig
+
+  if (minimalBuildNumber === maximalBuildNumber) return minimalBuildNumber === appBuildVersion
```
[the remark this answers is not in the document]
Comment (aliraiki) [reply src/libs/firebase/firestore/featureFlags/useFeatureFlag.ts:35]: Je vais le remplacer par ça pour gérer le cas où les deux sont undefined
```
if (minimalBuildNumber === undefined && maximalBuildNumber === undefined) return false
```

[the remark this answers is not in the document]
Comment (bruno-ebstein-pass-culture) [reply src/libs/firebase/firestore/featureFlags/useFeatureFlag.ts]: avec [la proposition précédente](https://github.com/pass-culture/pass-culture-app-native/pull/5950#discussion_r1490654053) ça serait aussi géré

**Review by bruno-ebstein-pass-culture — APPROVED** [review]

File: src/libs/firebase/firestore/featureFlags/useFeatureFlag.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (aliraiki) [reply src/libs/firebase/firestore/featureFlags/useFeatureFlag.ts]: Non ? (`!minimal` sera `true` et `!maximal` sera `true`, du coup on active le FF au lieu de le désactiver)

[the remark this answers is not in the document]
Comment (bruno-ebstein-pass-culture) [reply src/libs/firebase/firestore/featureFlags/useFeatureFlag.ts]: ooooh oui, pardon, bien vu
